wordpress-mobile/WordPress-Android
# Issue/5528 notification wait for update service

Fixes #5528 

## A note on Notes
The way the notification flow works is: if the note is not a big note (less than 8k or so IIRC) it travels within the android notification payload, so as soon as it is received it is saved to a local table and only then the notification is built and sent to the UI dashboard, so when the user taps on it we retrieve it from the local database and show it. If it didn't travel with the payload itself, we try to connect and download it from the REST api and save to the db as soon as we get the push. If this fails, then when the user taps on the notification the NotificationsDetailActivity is presented, it tries to fetch it from the DB but it doesn’t exist, and so there’s the error found.

On another thread, what happens is the notifications list is being regularly refreshed by the NotificationsUpdateService, which is started every time the app comes to foreground. If the NotificationsDetailActivity is presented before the update service finishes its task, then there is a possibility that this fails and produces this undesired behavior.


## The approach in this PR
With this PR what we’re doing is we check if we have the note saved in the db locally, and in case we can’t find it we start the update service trying to look for it and re-update.
This use case can also be triggered with the swipe to refresh pattern, so this way the user can recover from bad connectivity and retry on demand.
Finally, there is always a possibility that you tap on a notification for a Note that has been  deleted server side, and given this is a permanent error we show a toast and finish the detail activity as per what the original intention was.

Also, please note that the pull to refresh would preferrably load just the requested note, but given this works within a PagerAdapter structure (so you can swipe left-right to navigate the list horizontally) we are taking advantage of the fact that a connection needs to happen anyway (as the refresh needs to happen), and requesting the list of notes _as is_  in the `NotificationsUpdateService` and refreshing it completely (so to cope with the possibility some other notes have changed / been deleted in the meantime).

To test:
## TEST A
0. send the app to the background
1. have another user like/comment on a comment/post of yours
2. a notification should be received.
3. set your handset to offline
4. tap on the notification
5. the notification detail screen is shown (because it was saved when the push was received)
6. now pull to refresh
7. “There is no network available” toast is shown.

## TEST B
0. insert this code in line 852 in WordPress.java
```
                        // FIXME TEST CODE
                        NotificationsTable.deleteNoteById(getNoteIdFromNoteDetailActivityIntent(activity.getIntent()));
                        // FIXME END TEST CODE
                        // FIXME UNCOMMENT FOLLOWING LINES
//                        NotificationsUpdateService.startService(getContext(),
//                                getNoteIdFromNoteDetailActivityIntent(activity.getIntent()));
```

1. send the app to the background
2. have another user like/comment on a comment/post of yours
3. a notification should be received.
4. tap on the notification
5. the app tries to open the notification but deletes the local saved copy, so it starts trying to load it from the server
6. once loaded, you should see the notification

## TEST C
1. send the app to the background
2. have another user like/comment on a comment/post of yours
3. a notification should be received.
4. tap on the notification. You should see it, because it’s saved locally.
5. now delete the notification server-side (go to the web and trash it)
6. now pull to refresh
7. once refresh finishes, given the note is not found on the server, the error is shown and the screen is dismissed.

cc @daniloercoli @maxme
